sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SchedulerBuilder.java

```diff
@@ -0,0 +1,503 @@
+                } else {
+                    plans = Collections.emptyList();
+                }
+            } catch (ConfigStoreException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        LOGGER.info("Got {} {} plan{}: {}",
+                plans.size(),
+                plansType,
+                plans.size() == 1 ? "" : "s",
+                plans.stream().map(plan -> plan.getName()).collect(Collectors.toList()));
+        return plans;
+    }
+
+    /**
+     * Detects whether or not the previous deployment's type was set or not and if not, sets it in the state store.
+     *
+     * @param stateStore The stateStore to get last deployment type from.
+     * @param configStore The configStore to get plans from.
+     */
+    private static StateStoreUtils.DeploymentType getLastCompletedDeploymentType(
+            StateStore stateStore, Collection<Plan> oldPlans) {
+        StateStoreUtils.DeploymentType lastDeploymentType = StateStoreUtils.getLastCompletedUpdateType(stateStore);
+        if (lastDeploymentType.equals(StateStoreUtils.DeploymentType.NONE)) {
+            Optional<Plan> deployPlan = SchedulerUtils.getDeployPlan(oldPlans);
+            if (deployPlan.isPresent() && deployPlan.get().isComplete()) {
+                lastDeploymentType = StateStoreUtils.DeploymentType.DEPLOY;
+                StateStoreUtils.setLastCompletedUpdateType(stateStore, lastDeploymentType);
+            }
+        }
+        return lastDeploymentType;
+    }
+
+    /**
+     * Updates the Deploy plan in the provided list of {@code plans} to contain the provided {@code errors}.
+     * Returns a new list of plans containing the updates.
+     */
+    private static Collection<Plan> updateDeployPlan(Collection<Plan> plans, List<String> errors) {
+        Collection<Plan> updatedPlans = new ArrayList<>();
+        Plan deployPlan = SchedulerUtils.getDeployPlan(plans).get();
+        deployPlan = new DefaultPlan(
+                deployPlan.getName(),
+                deployPlan.getChildren(),
+                deployPlan.getStrategy(),
+                errors);
+
+        updatedPlans.add(deployPlan);
+        for (Plan plan : plans) {
+            if (!plan.isDeployPlan()) {
+                updatedPlans.add(plan);
+            }
+        }
+
+        return updatedPlans;
+    }
+
+    @VisibleForTesting
+    static ConfigStore<ServiceSpec> createConfigStore(
+            ServiceSpec serviceSpec, Collection<Class<?>> customDeserializationSubtypes, Persister persister)
+                    throws ConfigStoreException {
+        return new ConfigStore<>(
+                DefaultServiceSpec.getConfigurationFactory(serviceSpec, customDeserializationSubtypes),
+                persister);
+    }
+
+    /**
+     * Given the plans specified and the update scenario, the deploy plan may be overriden by a specified update plan.
+     */
+    @VisibleForTesting
+    static Collection<Plan> overrideDeployPlan(
```

Comment:
This also seems to affect the update plan. The name could be updated to indicate this.